### PR TITLE
Jumpdest cache bits

### LINKED ISSUE #2607: jump opcode analysis
### Description
Every time an executable account (contract) is run, a brand new Code object is instantiated from the bytecode in the account. Upon encountering the first JUMP instruction, all valid jump locations are computed and stored in a bitmap on the Code object. Unfortunately, that Code object is not retained for future use, so this analysis must be repeated on the first JUMP instruction of every contract, every time it is run.

### Acceptance Criteria

1. Reuse of jump destinations on all subsequent contract executions. 
2. Runtime speed improvements proportional to the number of jump dests in the contract. Pathological EVMtool test case [attached](https://gist.github.com/jflo/6f46fd44fd706c3f26deaacadf5b1a45), should speed up considerably.
3. Cache entries must be evicted on Account deletion.
4. Cache entries must be evicted on contract self-destruct.
5. Entries should expire out of cache in LRU order when total cache weight is hit.

### PR DESCRIPTION
**Pull request by jflo**
## PR description

Adds caching of valid jump destinations in a contract. This should reduce re-work previously required each time any function in the contract is called.

## Fixed Issue(s)

fixes #2607 

## Changelog

- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: evm/build.gradle
Comment (matkt) [inline evm/build.gradle:43]: I think we need to move the version to the version.gradle file

**Review by matkt — APPROVED** [review]
LGTM

File: CHANGELOG.md
Comment (shemnon) [inline CHANGELOG.md:7]: Nit: newline after changelist row before header.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java:45]: Nit: this touch will make "git spelunking" hard.  Please undo before submit.

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java
Comment (shemnon) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java:177]: Optional: A cache-miss (new code, maybe "0x02") and a `Mockito.verify(toRun, times(2)).calculateJumpDests();` would make it clear on reading that the verify doesn't reset between checks.

File: evm/src/main/java/org/hyperledger/besu/evm/Code.java
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/Code.java:58]: Static side effects are not appropriate for constructors.  If you need side effects do it in a factory method that calls the constructor and then does the side effects.

File: evm/src/main/java/org/hyperledger/besu/evm/worldstate/AbstractWorldUpdater.java
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/worldstate/AbstractWorldUpdater.java:103]: This purge is unneeded.  If accounts A,B,C,... share the same code Create A, Create B, Delete B, Create C, Delete C, etc.  Even though B and later are deleted A remains and would benefit from the cached jumpdests.  I think we should rely on the the cache's expiration policy instead of pre-deleting (unless we have reference counting handy).

**Review by shemnon — COMMENTED** [review]
I cannot approve this PR.  I have asked gently numerous times that the cache not be stored as a static field.  No other data structure in Besu does this. Dependency Injection is the design standard.

A better place would be to keep the cache *and the jump dest analysis logic* in the EVM class.  There is one EVM per hard form in the mainnet protocol spec, so they will not change that often during run time.

Putting this there would insulate us from future changes that are in progress.  One specifically is the EVM container format that will include a pre-computed jump table and will not have the code starting at the zero index of the container.  It also keeps us from hanging on to a static singleton.

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java:34]: nit: add space after cached

File: evm/src/main/java/org/hyperledger/besu/evm/worldstate/WorldView.java
Comment (macfarla) [inline evm/src/main/java/org/hyperledger/besu/evm/worldstate/WorldView.java:23]: another spurious change

**Review by macfarla — COMMENTED** [review]
couple of small comments

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java:2]: nit: new files should have * Copyright contributors to Hyperledger Besu  https://wiki.hyperledger.org/display/TSC/Copyright+and+License+Policy

**Comment by matkt** [comment]
> I cannot approve this PR. I have asked gently numerous times that the cache not be stored as a static field. No other data structure in Besu does this. Dependency Injection is the design standard.
> 
> A better place would be to keep the cache _and the jump dest analysis logic_ in the EVM class. There is one EVM per hard fork in the mainnet protocol spec, so they will not change that often during run time.
> 
> Putting this there would insulate us from future changes that are in progress. One specifically is the EVM container format that will include a pre-computed jump table and will not have the code starting at the zero index of the container. It also keeps us from hanging on to a static singleton.

IMO we must try to find a consensus on how to do this before merging this PR

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java
Comment (jflo) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java:177]: That would not pass though, since we are spying on the Code instance. A new code being analyzed would not cause toRun to be re-analyzed. If we want to test explicitly for cache hits/misses, I can add that, but this is really watching to see that for a given Code which contains the Jump operation, the analysis is only run once.

**Comment by jflo** [comment]
Closed in favor of https://github.com/hyperledger/besu/pull/2821

**Pull request closed without merging**
